python/pyarrow/_dataset.pyx

```diff
@@ -0,0 +1,1000 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: language_level = 3
+
+from __future__ import absolute_import
+
+from cython.operator cimport dereference as deref
+
+from pyarrow.lib cimport *
+from pyarrow.includes.libarrow_dataset cimport *
+from pyarrow.compat import frombytes, tobytes
+from pyarrow._fs cimport FileSystem, FileStats, Selector
+
+
+def _forbid_instantiation(klass, subclasses_instead=True):
+    msg = '{} is an abstract class thus cannot be initialized.'.format(
+        klass.__name__
+    )
+    if subclasses_instead:
+        subclasses = [cls.__name__ for cls in klass.__subclasses__]
+        msg += ' Use one of the subclasses instead: {}'.format(
+            ', '.join(subclasses)
+        )
+    raise TypeError(msg)
+
+
+cdef class FileFormat:
+
+    cdef:
+        shared_ptr[CFileFormat] wrapped
+        CFileFormat* format
+
+    def __init__(self):
+        _forbid_instantiation(self.__class__)
+
+    cdef void init(self, const shared_ptr[CFileFormat]& sp):
+        self.wrapped = sp
+        self.format = sp.get()
+
+    @staticmethod
+    cdef wrap(shared_ptr[CFileFormat]& sp):
+        cdef FileFormat self
+
+        typ = frombytes(sp.get().type_name())
+        if typ == 'parquet':
+            self = ParquetFileFormat.__new__(ParquetFileFormat)
+        else:
+            raise TypeError(typ)
+
+        self.init(sp)
+        return self
+
+    cdef inline shared_ptr[CFileFormat] unwrap(self):
+        return self.wrapped
+
+
+cdef class ParquetFileFormat(FileFormat):
+
+    def __init__(self):
+        self.init(shared_ptr[CFileFormat](new CParquetFileFormat()))
+
+
+cdef class PartitionScheme:
```

Comment:
Is it necessary to expose partition scheme as a class at all? I think it would suffice to have factories like `make_hive_partition_scheme()`

Reply:
> Is it necessary to expose partition scheme as a class at all? I think it would suffice to have factories like make_hive_partition_scheme()

What would a ``make_hive_partition_scheme()`` then return if the partition scheme object itself is not exposed?

Reply:
```py
ds.partition_scheme(pa.Schema schema, string flavor='hive') -> ds.PartitionScheme
```

Although I prefer to have not opaque return types, so I can distinguish between:
```py
ds.partition_scheme(schema)
ds.partition_scheme(schema, flavor='hive')
```

If we return the same class then I don't have the ability to inspect that they are differ.